File: cpp/3083_is_substring_present/is_substring_present.cpp
#include <cassert>
#include <string>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  bool isSubstringPresent(string s) {
    vector<int> h(26);
    for (int i = 0; i + 1 < s.size(); i++) {
      int x = s[i] - 'a';
      int y = s[i + 1] - 'a';
      h[x] |= 1 << y;
      if (h[y] >> x & 1) {
        return true;
      }
    }
    return false;
  }
};

int main() {
  vector<tuple<string, bool>> tests{
      {"leetcode", true},
      {"abcba", true},
      {"abcd", false},
  };

  for (auto &[s, ans] : tests) {
    assert(Solution().isSubstringPresent(s) == ans);
  }
}

File: rust/3083_is_substring_present/is_substring_present.rs
struct Solution;

impl Solution {
    pub fn is_substring_present(s: String) -> bool {
        let mut h = vec![0; 26];
        let s = s.as_bytes();
        for i in 0..s.len() - 1 {
            let x = (s[i] - b'a') as usize;
            let y = (s[i + 1] - b'a') as usize;
            h[x] |= 1 << y;
            if h[y] >> x & 1 == 1 {
                return true;
            }
        }
        false
    }
}

fn main() {
    let tests = vec![("leetcode", true), ("abcba", true), ("abcd", false)];

    for (s, ans) in tests {
        assert_eq!(Solution::is_substring_present(s.to_string()), ans);
    }
}
